File: Judge/알고리즘실습문제/3장/B.cpp
#include <iostream>
#include <vector>
#include <string>
#include <algorithm>
 
using namespace std;
 
int divide(string arr, int k) {
    if (arr.length() < k) {
        return 0;
    }
     
    vector<int> check(26, 0);
     
    for (char c : arr) {
        check[c - 'a']++;
    }
         
    for (int i = 0; i < arr.length(); i++) {
        if (check[arr[i] - 'a'] < k) {
            int mid = i;
            return max(divide(arr.substr(0, mid), k), divide(arr.substr(mid+1), k));
        }
    }
     
    return arr.length();
}
 
int solve(string arr, int k) {
    return divide(arr, k);
}
 
int main() {
    int T;
    cin >> T;
    cin.ignore();
    for (int i = 0; i < T; i++) {
        string s;
        getline(cin, s);
        int k;
        cin >> k;
        cin.ignore();
         
        int maxLen = solve(s, k);
        cout << maxLen << endl;
    }
    return 0;
}

File: Judge/알고리즘실습문제/3장/B.py
import sys
import collections
  
  
def divide(arr, k):
    if len(arr) < k:
        return 0
        
    freq = collections.Counter(arr)
   
    for i in range(len(arr)):
        if freq[arr[i]] < k:
            splitArr = arr.split(arr[i])
            return max(divide(a, k) for a in splitArr)
       
    return len(arr)
    
    
def solve(arr, k):
    return divide(arr,k)
   
   
   
def main():
    T = int(sys.stdin.readline())
    for _ in range(T):
        s = sys.stdin.readline().rstrip()
        k = int(sys.stdin.readline())
     
        maxLen = solve(s, k)
        print(maxLen)
         
main()